File: ts/tests/discord.test.ts
import { loadCases } from './helpers';
import { sendDiscordMessage, sendDiscordCard, _setSleepFn } from '../src/discord';

interface DiscordCases {
  send_message: {
    expected_url_pattern: string;
    expected_payload_fields: string[];
    expected_auth_header_prefix: string;
  };
  send_card: {
    expected_extra_field: string;
    description: string;
  };
  retry_after: {
    max_retries: number;
    cases: Array<{
      name: string;
      responses: Array<{
        status: number;
        retry_after?: string | null;
        body?: Record<string, unknown>;
      }>;
      expected_result?: Record<string, unknown>;
      expected_error_contains?: string;
      expected_attempts?: number;
    }>;
  };
}

const CASES = loadCases<DiscordCases>('discord_cases.json');

describe('Discord send_message', () => {
  it('sends message with correct URL and auth', async () => {
    const mockFetch = async (): Promise<Response> => {
      return {
        status: 200,
        ok: true,
        headers: new Headers(),
        text: async () => '{}',
        json: async () => ({ id: 'msg-discord-1' }),
      } as Response;
    };

    const result = await sendDiscordMessage('token123', 'channel-id', 'Hello', mockFetch);
    expect(result).toEqual({ id: 'msg-discord-1' });
  });
});

describe('Discord send_card', () => {
  it('sends card with embeds field', async () => {
    const mockFetch = async (): Promise<Response> => {
      return {
        status: 200,
        ok: true,
        headers: new Headers(),
        text: async () => '{}',
        json: async () => ({ id: 'msg-discord-1' }),
      } as Response;
    };

    const card = { title: 'Test', description: 'Test embed' };
    const result = await sendDiscordCard('token123', 'channel-id', 'Hello', card, mockFetch);
    expect(result).toEqual({ id: 'msg-discord-1' });
  });
});

describe('Discord retry-after handling', () => {
  it.each(CASES.retry_after.cases)('$name', async (tc) => {
    let attemptCount = 0;
    const sleepCalls: number[] = [];

    // Mock sleep function to track calls
    _setSleepFn(async (ms: number) => {
      sleepCalls.push(ms / 1000); // Convert back to seconds
    });

    // Create a fetchFn that returns responses in sequence
    const fetchFn = async (): Promise<Response> => {
      if (attemptCount >= tc.responses.length) {
        throw new Error(`Unexpected fetch call ${attemptCount}`);
      }

      const response = tc.responses[attemptCount];
      attemptCount++;

      // Create a response-like object
      const headers: Record<string, string> = {};
      if (response.retry_after !== undefined && response.retry_after !== null) {
        headers['Retry-After'] = response.retry_after;
      }

      return {
        status: response.status,
        ok: response.status >= 200 && response.status < 300,
        headers: new Headers(headers),
        text: async () => JSON.stringify({ status: response.status }),
        json: async () => response.body || {},
      } as Response;
    };

    // Test the expected outcome
    if (tc.expected_error_contains) {
      await expect(sendDiscordMessage('token', 'channel-id', 'test', fetchFn)).rejects.toThrow(
        new RegExp(tc.expected_error_contains),
      );
    } else {
      const result = await sendDiscordMessage('token', 'channel-id', 'test', fetchFn);
      expect(result).toEqual(tc.expected_result);
    }

    // Verify attempt count
    if (tc.expected_attempts !== undefined) {
      expect(attemptCount).toBe(tc.expected_attempts);
    }
  });
});

File: go/discord_test.go
package wgrok

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"sync"
	"testing"
)

type discordTestCases struct {
	SendMessage struct {
		ExpectedURLPattern    string   `json:"expected_url_pattern"`
		ExpectedPayloadFields []string `json:"expected_payload_fields"`
		ExpectedAuthPrefix    string   `json:"expected_auth_header_prefix"`
	} `json:"send_message"`
	SendCard struct {
		ExpectedExtraField string `json:"expected_extra_field"`
	} `json:"send_card"`
	RetryAfter struct {
		MaxRetries int `json:"max_retries"`
		Cases      []struct {
			Name      string `json:"name"`
			Responses []struct {
				Status     int                    `json:"status"`
				RetryAfter interface{}            `json:"retry_after"`
				Body       map[string]interface{} `json:"body"`
			} `json:"responses"`
			ExpectedResult   map[string]interface{} `json:"expected_result"`
			ExpectedAttempts int                    `json:"expected_attempts"`
			ExpectedError    string                 `json:"expected_error_contains"`
		} `json:"cases"`
	} `json:"retry_after"`
}

func loadDiscordTestCases(t *testing.T) discordTestCases {
	t.Helper()
	data, err := os.ReadFile("../tests/discord_cases.json")
	if err != nil {
		t.Fatalf("load discord cases: %v", err)
	}
	var cases discordTestCases
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("parse discord cases: %v", err)
	}
	return cases
}

// We need to override the URL builder - create a test variable for it
var testDiscordChannelMessagesURL func(channelID string) string

func overrideDiscordChannelMessagesURL(t *testing.T, urlFunc func(channelID string) string) {
	t.Helper()
	orig := DiscordChannelMessagesURL
	DiscordChannelMessagesURL = urlFunc
	t.Cleanup(func() { DiscordChannelMessagesURL = orig })
}

func TestSendDiscordMessageHTTP(t *testing.T) {
	var capturedBody map[string]interface{}
	var capturedAuth string

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedAuth = r.Header.Get("Authorization")
		body, _ := io.ReadAll(r.Body)
		json.Unmarshal(body, &capturedBody)
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"id":"msg-discord-1","content":"hello"}`))
	}))
	defer srv.Close()

	// Override URL builder to use test server
	overrideDiscordChannelMessagesURL(t, func(channelID string) string {
		return srv.URL
	})

	result, err := SendDiscordMessage("bot-token-123", "123456", "hello", srv.Client())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["id"] != "msg-discord-1" {
		t.Errorf("result id = %v, want msg-discord-1", result["id"])
	}
	if capturedAuth != "Bot bot-token-123" {
		t.Errorf("auth = %q, want %q", capturedAuth, "Bot bot-token-123")
	}
	if capturedBody["content"] != "hello" {
		t.Errorf("content = %v, want hello", capturedBody["content"])
	}
}

func TestSendDiscordMessageHTTPError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(401)
		w.Write([]byte(`{"message":"Unauthorized"}`))
	}))
	defer srv.Close()

	overrideDiscordChannelMessagesURL(t, func(channelID string) string {
		return srv.URL
	})

	_, err := SendDiscordMessage("badtoken", "123456", "hello", srv.Client())
	if err == nil {
		t.Fatal("expected error, got nil")
	}
}

func TestSendDiscordCardHTTP(t *testing.T) {
	var capturedBody map[string]interface{}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		json.Unmarshal(body, &capturedBody)
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"id":"msg-embed-1"}`))
	}))
	defer srv.Close()

	overrideDiscordChannelMessagesURL(t, func(channelID string) string {
		return srv.URL
	})

	embed := map[string]interface{}{"title": "Card Title", "description": "Card description"}
	result, err := SendDiscordCard("tok", "123456", "fallback", embed, srv.Client())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["id"] != "msg-embed-1" {
		t.Errorf("result id = %v, want msg-embed-1", result["id"])
	}

	embeds, ok := capturedBody["embeds"].([]interface{})
	if !ok || len(embeds) != 1 {
		t.Fatalf("expected 1 embed, got %v", capturedBody["embeds"])
	}
	embedMap := embeds[0].(map[string]interface{})
	if embedMap["title"] != "Card Title" {
		t.Errorf("embed title = %v, want Card Title", embedMap["title"])
	}
}

func TestDiscordRetryAfter(t *testing.T) {
	cases := loadDiscordTestCases(t)
	for _, tc := range cases.RetryAfter.Cases {
		t.Run(tc.Name, func(t *testing.T) {
			responseIndex := 0
			var mu sync.Mutex

			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				mu.Lock()
				defer mu.Unlock()

				if responseIndex >= len(tc.Responses) {
					t.Errorf("unexpected request: response index %d >= %d", responseIndex, len(tc.Responses))
					w.WriteHeader(500)
					return
				}

				resp := tc.Responses[responseIndex]
				responseIndex++

				w.Header().Set("Content-Type", "application/json")
				if resp.RetryAfter != nil {
					w.Header().Set("Retry-After", resp.RetryAfter.(string))
				}
				w.WriteHeader(resp.Status)

				if resp.Body != nil {
					body, _ := json.Marshal(resp.Body)
					w.Write(body)
				} else {
					w.Write([]byte("{}"))
				}
			}))
			defer srv.Close()

			overrideDiscordChannelMessagesURL(t, func(channelID string) string {
				return srv.URL
			})

			result, err := SendDiscordMessage("tok", "123456", "test", srv.Client())

			if tc.ExpectedResult != nil {
				if err != nil {
					t.Errorf("expected success, got error: %v", err)
				}
				if result == nil || result["id"] != tc.ExpectedResult["id"] {
					t.Errorf("result id = %v, want %v", result["id"], tc.ExpectedResult["id"])
				}
			}

			if tc.ExpectedError != "" {
				if err == nil {
					t.Errorf("expected error containing %q, got nil", tc.ExpectedError)
				} else if !strings.Contains(err.Error(), tc.ExpectedError) {
					t.Errorf("error = %v, want to contain %q", err, tc.ExpectedError)
				}
			}

			if tc.ExpectedAttempts > 0 {
				if responseIndex != tc.ExpectedAttempts {
					t.Errorf("expected %d attempts, got %d", tc.ExpectedAttempts, responseIndex)
				}
			}
		})
	}
}
